torch/nn/parallel/data_parallel.py

```diff
@@ -1,10 +1,67 @@
+import weakref
 import torch
+import torch.cuda.comm as comm
+from collections import OrderedDict
 from ..modules import Module
 from .scatter_gather import scatter_kwargs, gather
 from .replicate import replicate
 from .parallel_apply import parallel_apply
 
 
+# NOTE: these callbacks are one-off
+# NOTE: they are dedupliacted based on identity
+class _CallbackOrderedDict(OrderedDict):
+    @classmethod
+    def _init_class(cls):
+        mutating_methods = [
+            '__setitem__',
+            '__delitem__',
+            'pop',
+            'popitem',
+            'update',
+            'setdefault'
+        ]
+
+        def method_wrapper(name):
+            def callback_method(self, *args, **kwargs):
+                for callback in self.callbacks:
+                    callback()
+                del self.callbacks[:]
+                return getattr(super(_CallbackOrderedDict, self), name)(*args, **kwargs)
+            return callback_method
+        for name in mutating_methods:
+            setattr(cls, name, method_wrapper(name))
+
+    @classmethod
+    def from_ordered(cls, od):
+        return cls(od.items())
+
+    def __init__(self, *args, **kwargs):
+        self.callbacks = []
+        super(_CallbackOrderedDict, self).__init__(*args, **kwargs)
+
+    def register_modification_callback(self, cb):
+        if all(c is not cb for c in self.callbacks):
```

Comment:
Quadratic police! If you *really really really* didn't want callbacks to show up multiple times in the callback list (I don't really know why this is necessary), couldn't you store them in a `set()`? Or an OrderedDict even, if you wanted to preserve callback order as the order they were added.

Reply:
Yes I really really really don't want it, because if one module changes, only its callback will get cleared, and I don't want to register them again in **all modules** the next time we initialize the cache (this would give us quadratic run time given that you modify all modules!).

The complexity of this line doesn't matter. Right now the number of callbacks will be equal to number of `DataParallel` modules a given submodule is nested in, and I don't know of any cases where people use more than 1. I can add a warning that you shouldn't add too many of them or you should fix this method if you'd prefer that.

Reply:
> The complexity of this line doesn't matter. Right now the number of callbacks will be equal to number of DataParallel modules a given submodule is nested in, and I don't know of any cases where people use more than 1.

If that is the assumption being relied on, then it should be in the code ;)  It's especially poignant here because the class looks like it should be general purpose way to have OrderedDict with callbacks, but as written is only a good idea when you don't put too many callbacks on it. And leave it to end users to misuse code for things they weren't designed for ;)